readme.md
# whyplyaer
一个开源的html5的播放器，使用方式很简单。使用自定义播放进度条来统一各个平台上播放器样式。播放器以16：9的比例显示，横竖屏下都保持此比例。

本项目包含两种实现方式，一种使用 jQuery 实现，一种使用 [web components v1](https://developers.google.com/web/fundamentals/web-components/customelements) 实现(Chrome 需要 54+)。

## 示例

### 使用 jQuery
在使用前要引入 jQuery，同时需要引入 whyun player 的 css 和 js 文件，具体代码可以参见[demo文件](https://github.com/yunnysunny/whyun-player/blob/master/demo/jquery/demo.html)。效果参见 [在线演示地址](http://silian.whyun.com/whyun-player/demo.html)。
### 使用 web components
直接引用 whyun player 的 js 代码即可，其内部会引用所需的 css 文件，具体代码可以参见[这里](https://github.com/yunnysunny/whyun-player/blob/master/demo/webcomponent/demo.html)。

> 不管使用以上哪种方式，如果想支持 PC 上播放 m3u8 ，手动引入 [hls.js](https://github.com/video-dev/hls.js/) 即可。

## API

### jQuery API

#### $.whyun.player(options)

**参数**

- options.parent {String|Dom|jQuery} 当前播放器要展现的父层容器，可以传 dom 选择器、dom 对象或者 jquery 对象
- options.src {String} 当前播放器的视频播放地址
- options.poster {String} 当前播放器的海报地址
- options.debug {Boolean} 是否打印调试日志

#### $.whyun.player.setSrc

**参数**

- src {String} 要更换的播放地址


#### $.whyun.player.setPoster

- poster {String} 要更换的海报地址

### web components API

```html
<whyun-player debug src="" poster=""></whyun-player>
```
由于是使用 web components ，所以属性值可以直接在 DOM 节点上赋值，手动修改节点属性或者在 js 对象中给 `src` `poster` 赋值，都可以立即生效。例如有如下代码：
```html
<whyun-player debug src="原始播放地址" poster="" id="myplayer"></whyun-player>
```
```javascript
document.getElementById('myplayer').src = '要更改的播放地址';//调用完之后可以立即换播放地址
```

## 维护者
[yunnysunny](https://github.com/yunnysunny)

## License

MIT

changelog.md
# changelog
## 0.2.0

1. 增加了 web components 方式的 API。

## 0.1.0

1. 增加在 PC 上播放 m3u8 的功能。

## 0.0.2
1. 优化了进度条的布局
## 0.0.1
1. 项目初始化发布

electron/app.js
const electron = require('electron');
// const fs = require('fs');
// const path = require('path');
const globalShortcut = electron.globalShortcut;
// Module to control application life.
const app = electron.app;
// Module to create native browser window.
const BrowserWindow = electron.BrowserWindow;

// Keep a global reference of the window object, if you don't, the window will
// be closed automatically when the JavaScript object is garbage collected.
let mainWindow;

global.releaseVersion = false;

function createWindow() {
    // 创建窗体
    mainWindow = new BrowserWindow({
        width: 1280, height: 720,
        webPreferences: {
            tabbingIdentifier: 'groups',
            webSecurity: false,
            nodeIntegration: false,
            // preload: path.join(__dirname, 'interface.js'),
            plugins: true
        }
    });

    // 注册打开控制台的快捷键
    globalShortcut.register('CommandOrControl+P+0', function () {
        mainWindow.webContents.openDevTools();
    });

    // 加载首页
    mainWindow.loadURL('file://' + __dirname + '/../demo/webcomponent/demo.html');

    // 创建窗体时打开控制台
    //   mainWindow.webContents.openDevTools()

    // 监听关闭窗体事件
    mainWindow.on('closed', function () {
        mainWindow = null;
    });


}

app.on('ready', createWindow);

app.on('window-all-closed', function () {
    if (process.platform !== 'darwin') {
        app.quit();
    }
});

app.on('activate', function () {
    if (mainWindow === null) {
        createWindow();
    }
});


lib/jquery/jquery.whyun.player.js
/**
* @lisence MIT
* @copyright yunnysunny<yunnysunny@gmail.com>
*/
(function ($) {
    function addZero(time) {
        time = parseInt(time, 10);
        return time < 10 ? '0' + time : time;
    }
    function secondsToString(seconds) {
        if (!seconds || seconds === Infinity) {
            return '00:00';
        }
        var h = Math.floor(seconds / 3600);
        var left = seconds % 3600;
        var m = Math.floor(left / 60);
        var s = seconds % 60;
        if (h === 0) {
            return addZero(m) + ':' + addZero(s);
        }
        return addZero(h) + ':'  +addZero(m) + ':' + addZero(s);
    }
    function getPageX(e) {
        if (e.pageX) {
            return e.pageX;
        }
        var event = e.originalEvent;
        if (!event.changedTouches || event.changedTouches.length === 0) {
            return 0;
        }
        var touch = event.changedTouches[event.changedTouches.length - 1];
        return touch.pageX;
    }
    function parseUrl(url) {
        try {
            url = new URL(url);
        } catch(e) {
            //
        }
        
        return url;
    }
    /**
     * 
     * @param {Object} options 
     * @param {String|Dom|jQuery} options.parent 当前播放器要展现的父层容器，可以传dom选择器、dom对象或者jquery对象
     * @param {String} options.src  当前播放器的视频播放地址
     * @param {String} options.poster  当前播放器的海报地址
     * @param {Boolean} options.debug 是否打印调试日志
     */
    function WhyunPlayer(options) {
        var parent = options.parent;
        if (typeof (parent) == 'string') {
            parent = $(parent);
        } else if (parent.nodeType === 1) {
            parent = $(parent);
        }
        this.parent = parent;
        this.src = options.src || '';
        this._videoType = options.videoType;

        this.poster = options.poster || '';
        this.width = options.width || '100%';
        this._debug = options.debug;

        this.id = ('player' + Math.random()).replace('.', '');
        this.attrs = options.attrs || {};
        var canvas = options.canvas;
        if (canvas) {
            this.canvas = canvas;
        }

        var ua = window.navigator.userAgent.toLowerCase();
        this.isIos = (/(iphone|ipad|ipod)/i).test(ua);
        this._isAndroid = (/android/i).test(ua);
        this._isMobile = this.isIos || this._isAndroid;
        this.canUseNativeFullApi = (/(qqbrowser|ucbrowser|micromessenger)/i).test(ua);
        this._shouldUseHlsJsWhenPlayM3u8 = !this._isMobile && window.Hls && Hls.isSupported();
        this._initDom();
        this._addEvent();
    }
    WhyunPlayer.prototype._debugLog = function() {
        if (this._debug) {
            // eslint-disable-next-line no-console
            console.info.apply(console, arguments);
        }
    };
    WhyunPlayer.prototype._isM3u8 = function() {
        var urlObj = parseUrl(this.src) || {};
        var path = (urlObj.pathname || '');
        var isM3u8Now = false;
        if (path.indexOf('.m3u8') !== -1) {
            isM3u8Now = true;
        }
        return isM3u8Now;
    };
    WhyunPlayer.prototype._renderVideo = function() {
        var hls = null;
        var src = this.src;
        var useHlsJs = false;
        var $video = this.$video;
        var video = this._videoDom;

        if (this._isM3u8() && this._shouldUseHlsJsWhenPlayM3u8) {
            useHlsJs = true;
            if (hls) {
                hls.destroy();
            }
            hls = new Hls();
            hls.loadSource(src);
            hls.attachMedia(video);
            hls.on(Hls.Events.MANIFEST_PARSED,function() {
                video.play();
            });
        }
        if (!useHlsJs) {
            $video.children('source').attr('src',src);
        }
    };

    WhyunPlayer.prototype._initDom = function () {
        var $parent = this.parent;
        $parent.empty().addClass('video-container');
        var $video = $('<video><source /></video>');
        var attrs = {
            'class': '',
            id: this.id,
            perload:'auto',
            //            autoplay:true,
            'x-webkit-airplay':'allow',
            'webkit-playsinline':true,
            'x5-video-player-type' : 'h5',

            poster: this.poster
        };
        $.extend(attrs,this.attrs);

        $video.attr(attrs);//视频

        $parent.append($video);
        this._videoDom = $video.get(0);
        this.$video = $video;
        this._renderVideo();

        if (this.canvas) {
            var $canvas = $('<canvas></canvas>');
            $parent.append($canvas);
            this.$canvas = $canvas;
            //$video.css({'object-position':'0px 0px'});
        }
        

        if (this.poster) {//海报
            var $poster = $('<img />');
            $poster.attr({
                src:this.poster
            });
            $parent.append($poster);
            $video.hide();
            this.$poster = $poster;
        } else {
            this.$poster = $('none');
        }

        var $playBtn = $('<div class="play-button show"><div class="triangle"></div></div>');
        var $loadingContainer = $('<div class="uil-default-css" style="-webkit-transform:scale(0.4)"></div>');
        for (var i=0;i<12;i++) {
            var n = i * 30;
            $loadingContainer.append($('<div></div>').css({
                '-webkit-transform':'rotate('+n+'deg) translate(0,-60px)',
                transform:'rotate('+n+'deg) translate(0,-60px)'
            }));
        }

        $playBtn.append($loadingContainer);
        $parent.append($playBtn);
        this.$playBtn = $playBtn;

        var $controller = $(this._getControllerHtml());
        $parent.append($controller);
        //$controller.find('.progress-controls').width($controller.width() - $controller.find('.left-controls').width() - $controller.find('.function-controls').width());
        this.$controller = $controller;
    };
    WhyunPlayer.prototype._getControllerHtml = function() {
        //        var player = this._videoDom;
        var html = '<div class="play-controls">' +
            '<div class="left-controls">' +
                '<div class="play-control"></div>' +
            '</div>' +
            '<div class="progress-controls">' +
                '<div class="progress">' +
                    '<div class="all-len"></div>'+
                    '<div class="played"><div class="track" draggable="true"></div></div>' +
                    '<div class="loaded"></div>' +
                '</div>' +

            '</div>' +
            '<div class="function-controls">' +
                '<div class="time-wrap">' +
                    '<span class="time-passed">00:00</span><span>/</span><span class="time-total">00:00</span>' +
                '</div>' +

                '<div class="fullscreen-wrap"><div class="fullscreen-control"></div></div>' +
        //                    '<div class="sound"><a href="#sound">mute</a></div>' +
            '</div>' +
            '</div>';
        return html;
    };
    WhyunPlayer.prototype._addEvent = function () {
        var _self = this;
        var $video = this.$video;
        var player = this._videoDom;
        var $poster = _self.$poster;
        var $playBtn = _self.$playBtn;
        var $controller = _self.$controller;
        var $leftControlArea = $controller.find('.left-controls');
        var $playControl = $leftControlArea.find('.play-control');
        var $played = $controller.find('.played');
        var $progress = $played.parent();
        var $progressController = $progress.parent();
        var $track = $controller.find('.track');
        var $loaded = $controller.find('.loaded');
        var $fullscreenArea = $controller.find('.fullscreen-wrap');

        var $timePassed = $controller.find('.time-passed');
        var $timeTotal = $controller.find('.time-total');
        var INTERVAL_HIDE_CONTROLLER = 10 * 1000;
        var hideControllerTimer = null;
        var trackWidth = $track.width();
        
        var canvas =  null;
        if (this.canvas) {
            canvas = this.$canvas.get(0).getContext('2d');
            player.style['object-position']= '0px 0px';
        }

        function showAndHideController() {
            if (hideControllerTimer) {
                clearTimeout(hideControllerTimer);
            }
            $controller.show();
            hideControllerTimer = setTimeout(function() {
                $controller.hide();
            },INTERVAL_HIDE_CONTROLLER);
        }
        showAndHideController();

        function pauseShow() {
            //            $video.hide();
            //            $poster.show();
            $playBtn.removeClass('loading').show();
            $playControl.removeClass('pause');
        }
        function playShow() {
            $poster.hide();
            $playBtn.addClass('loading');
            if (!this.canvas) {
                $video.show();
            }
            
            player.play();
            $playControl.addClass('pause');
        }

        function getX(e) {
            var pageX = getPageX(e);
            if (!pageX) {
                return -1;
            }
            var parentOffsetLeft = $progress.offset().left;
            //or $(this).offset(); if you really just want the current element's offset
            var relX = pageX - parentOffsetLeft;
            _self._debugLog(e.type,relX);
            return relX;
        }

        function playToGivenPosition(e) {
            if (!player.duration) {
                return;
            }
            var relX = getX(e);
            if (relX === -1) {
                return;
            }
            var percentage = relX / $progress.width();
            player.currentTime =( player.duration * percentage);
            $played.width((percentage * 100) + '%' );
            _self._debugLog(relX, $progressController.width(), percentage, player.duration, player.currentTime);
            // player.play();
        }
        function showLoadedPosition() {
            var buffered = player.buffered;
            if (buffered && buffered.length > 0) {
                var buf = buffered.end(buffered.length - 1);
                var percentage = buf / player.duration;
                if ($progress.width() * percentage > trackWidth) {
                    $loaded.width(percentage * 100 + '%');
                } else {
                    $loaded.width(trackWidth);
                }
            }
        }
        function draw() {
            if(player.paused || player.ended) {
                return false;
            } 
            // First, draw it into the backing canvas
            canvas.drawImage(player,0,0,player.clientWidth,player.clientHeight);
            // Start over!
            setTimeout(function(){ draw(); }, 0);
        }

        $playBtn.click(function() {
            playShow();
        });
        $leftControlArea.click(function() {
            if ($playControl.hasClass('pause')) {//当前操作为暂停
                player.pause();
            } else {//当前操作为播放
                playShow();
            }
        });


        player.addEventListener('error', function() {
            _self._debugLog('emit error events');
            pauseShow();
        }, false);


        player.addEventListener('pause', function() {
            _self._debugLog('emit pause events');
            pauseShow();
        }, false);

        player.addEventListener('timeupdate', function() {
            var currentTime = player.currentTime;
            if (!currentTime) {
                return;
            }
            $poster.hide();
            $playBtn.hide();
            $playControl.addClass('pause');
            if (!_self.canvas) {
                $video.show();
            }
            

            $timePassed.text(secondsToString(currentTime));
            var percentage =( currentTime  / player.duration);

            if ($progress.width() * percentage > trackWidth) {
                $played.width((percentage * 100) + '%');
            } else {
                $played.width(trackWidth);
            }
        }, false);
        player.addEventListener('loadedmetadata',function() {_self._debugLog(player.videoWidth);
            if (player.duration) {
                $timeTotal.text(secondsToString(player.duration));
            }
        });
        player.addEventListener('durationchange', function() {
            if (player.duration) {
                $timeTotal.text(secondsToString(player.duration));
            }
        }, false);
        player.addEventListener('loadeddata',function() {
            showLoadedPosition();
        },false);
        player.addEventListener('canplaythrough',function() {
            showLoadedPosition();
        },false);
        player.addEventListener('progress',function() {
            showLoadedPosition();
        });
        if (this.canvas) {
            player.addEventListener('play',function() {
                draw();
            },false);
        }
        
        $progress.click(function(e){//进度条点击
            playToGivenPosition(e);
        });
        $progress.on('mousedown touchstart',function(e) {
            showAndHideController();
            e.preventDefault();
        });
        $progress.on('mouseup touchend',function(e) {
            playToGivenPosition(e);
        });
        this.parent.on('mouseenter touchstart',function() {//触摸播放器，显示进度条
            showAndHideController();_self._debugLog('player touch');
        });

        $(window).resize(function () {
        });
        $fullscreenArea.click(function() {
            _self._toggleFullScreen();
        });
    };
    WhyunPlayer.prototype._toggleFullScreen = function () {
        if (this.canUseNativeFullApi) {
            this._debugLog('use native full api');
            var elem = this._videoDom;

            if (elem.requestFullscreen) {
                this._debugLog('requestFullscreen');
                elem.requestFullscreen();
            } else if (elem.mozRequestFullScreen) {
                this._debugLog('requestFullscreen2');
                elem.mozRequestFullScreen();
            } else if (elem.webkitRequestFullscreen) {
                this._debugLog('requestFullscreen3');
                elem.webkitRequestFullscreen();
            } else if (elem.webkitEnterFullscreen) {
                this._debugLog('webkitEnterFullscreen');
                elem.webkitEnterFullscreen();
            } else {
                this._debugLog('not support full api');
            }
            return;
        }
        var $html = $('html');this._debugLog('use custom full api');
        $html.toggleClass('fullscreen');
    };

    /**
     * @param {String} src 要设置的视频播放地址
     */
    WhyunPlayer.prototype.setSrc = function (src) {
        this.src = src;
        var player = this._videoDom;
        player.pause();
        this.$playBtn.addClass('loading').show();
        this.$video.children('source').attr('src', src);
        player.currentTime = 0;
        this._renderVideo();
        return this;
    };
    /**
     * @param {String} poster 要设置的海报地址
     */
    WhyunPlayer.prototype.setPoster = function(poster) {
        this.$poster.attr('src', poster);
        this.$video.attr('poster', poster);
        return this;
    };
    $.whyun = $.extend($.whyun,{});
    $.whyun.player = function(option) {
        return new WhyunPlayer(option);
    };
})(jQuery);

lib/webcomponent/whyun_player.js
(function () {
    function addZero(time) {
        time = parseInt(time, 10);
        return time < 10 ? '0' + time : time;
    }
    function secondsToString(seconds) {
        if (!seconds || seconds === Infinity) {
            return '00:00';
        }
        var h = Math.floor(seconds / 3600);
        var left = seconds % 3600;
        var m = Math.floor(left / 60);
        var s = seconds % 60;
        if (h === 0) {
            return addZero(m) + ':' + addZero(s);
        }
        return addZero(h) + ':' + addZero(m) + ':' + addZero(s);
    }
    function getPageX(e) {
        if (e.pageX) {
            return e.pageX;
        }
        var event = e.originalEvent;
        if (!event.changedTouches || event.changedTouches.length === 0) {
            return 0;
        }
        var touch = event.changedTouches[event.changedTouches.length - 1];
        return touch.pageX;
    }
    function parseUrl(url) {
        try {
            url = new URL(url);
        } catch (e) {
            //
        }

        return url;
    }
    function addListenerMulti(element, eventNames, listener) {
        var events = eventNames.split(' ');
        for (var i = 0, iLen = events.length; i < iLen; i++) {
            element.addEventListener(events[i], listener, false);
        }
    }
    function setWidth(dom, width) {
        var value = '';
        if (typeof (width) === 'number') {
            value = width + 'px';
        } else if (typeof (width) === 'string') {
            if (isNaN(Number(width))) {
                value = width;
            } else {
                value = width + 'px';
            }
        } else {
            // eslint-disable-next-line no-console
            console.error('错误的宽度', width, dom);
        }
        return dom.style.width = value;
    }
    function getStyle(el, styleProp) {
        var defaultView = (el.ownerDocument || document).defaultView;
        // W3C standard way:
        // sanitize property name to css notation
        // (hypen separated words eg. font-Size)
        styleProp = styleProp.replace(/([A-Z])/g, '-$1').toLowerCase();
        return defaultView.getComputedStyle(el, null).getPropertyValue(styleProp);
       
    }
    HTMLElement.prototype.show = function (display) {
        this.style.display = display || '';
    };
    HTMLElement.prototype.hide = function () {
        this.style.display = 'none';
    };
    const template = `
    <link rel="stylesheet" type="text/css" href="../../css/jquery.ohwit.player.css" />
    <div class="video-container">
        
        <video x-webkit-airplay webkit-playsinline x5-video-player-type="h5" perload="auto">
            <source />
        </video>
        <div class="play-button show" id="js-play-button">
            <div class="triangle"></div>
            <div class="uil-default-css" style="transform:scale(0.4)"></div>
        </div>
        <div class="play-controls" id="js-play-controls">
            <div class="left-controls">
                <div class="play-control"></div>
            </div>
            <div class="progress-controls">
                <div class="progress">
                    <div class="all-len"></div>
                    <div class="played"><div class="track" draggable="true"></div></div>
                    <div class="loaded"></div>
                </div>

            </div>
            <div class="function-controls">
                <div class="time-wrap">
                    <span class="time-passed">00:00</span><span>/</span><span class="time-total">00:00</span>
                </div>
                <div class="fullscreen-wrap">
                    <div class="fullscreen-control"></div>
                </div>
                <!--div class="sound"><a href="#sound">mute</a></div-->
            </div>
        </div>
    </div>
    `;
    /**
     * @class WhyunPlayer
     * @extends HTMLElement
     */
    class WhyunPlayer extends HTMLElement {
        constructor() {
            super();

            const root = this.attachShadow({ mode: 'open'/*, delegatesFocus: true*/ });
            root.innerHTML = template;
            this._root = root;
            this.src = this.getAttribute('src');
            this.poster = this.getAttribute('poster') || '';
            this._debug = this.hasAttribute('debug') || false;

            var ua = window.navigator.userAgent.toLowerCase();
            var _isIos = (/(iphone|ipad|ipod)/i).test(ua);
            var _isAndroid = (/android/i).test(ua);
            this._isMobile = _isIos || _isAndroid;
            this._canUseNativeFullApi = (/(qqbrowser|ucbrowser|micromessenger)/i).test(ua);
            this._shouldUseHlsJsWhenPlayM3u8 = !this._isMobile && window.Hls && Hls.isSupported();
            this._hls = null;
        }

        _isM3u8(link) {
            var urlObj = parseUrl(link) || {};
            var path = (urlObj.pathname || '');
            var isM3u8Now = false;
            if (path.indexOf('.m3u8') !== -1) {
                isM3u8Now = true;
            }
            return isM3u8Now;
        }
        /**
         * @param {String} link
         */
        set src(link) {
            if (this._isM3u8(link) && this._shouldUseHlsJsWhenPlayM3u8) {
                let hls = this._hls;
                if (hls) {
                    hls.destroy();
                }
                this._hls = hls = new Hls();
                hls.loadSource(link);
                const video = this._root.querySelector('video');
                hls.attachMedia(video);
                hls.on(Hls.Events.MANIFEST_PARSED,function() {
                    video.play();
                });
            } else {
                this._root.querySelector('source').setAttribute('src', link);
            }

            
        }
        /**
         * @param {String} link
         */
        set poster(link) {
            const video = this._root.querySelector('video');
            if (!link) {
                return video.removeAttribute('poster');
            }
            video.setAttribute('poster', link);
        }

        connectedCallback() {
            //元素每次插入到 DOM 时都会调用。用于运行安装代码，例如获取资源或渲染。一般来说，您应将工作延迟至合适时机执行。
            // console.log('connectedCallback');
            const _this = this;
            this._root.querySelector('link').addEventListener('load',function() {
                _this._addEvent();
            }, false);
            
        }
        static get observedAttributes() {
            return ['src', 'poster'];
        }
        
        attributeChangedCallback(attrName, oldVal, newVal) {
            //如果有人对您的元素调用 el.setAttribute(...)，浏览器将立即调用 attributeChangedCallback()。
            if (attrName === 'src') {
                this.src = newVal;
            } else if (attrName === 'poster') {
                this.poster = newVal;
            }
            this._debugLog(attrName, oldVal, newVal);
        }
        disconnectedCallback() {
            // 从 DOM 中移除元素（例如用户调用 el.remove()）后，您会立即收到 disconnectedCallback()
        }
        _debugLog() {
            if (this._debug) {
                // eslint-disable-next-line no-console
                console.info.apply(console, arguments);
            }
        }
        _addEvent() {
            var _self = this;
            var root = this._root;
            var player = root.querySelector('video');
            // var $poster = _self.$poster;
            var playBtn = root.querySelector('#js-play-button');
            var $controller = root.querySelector('#js-play-controls');
            var $leftControlArea = $controller.querySelector('.left-controls');
            var $playControl = $leftControlArea.querySelector('.play-control');
            var $played = $controller.querySelector('.played');
            var $progress = $controller.querySelector('.progress');
            // var $progressController = $controller.querySelector('.progress-controls');
            var $track = $controller.querySelector('.track');
            var $loaded = $controller.querySelector('.loaded');
            var $fullscreenArea = $controller.querySelector('.fullscreen-wrap');

            var $timePassed = $controller.querySelector('.time-passed');
            var $timeTotal = $controller.querySelector('.time-total');
            var INTERVAL_HIDE_CONTROLLER = 10 * 1000;
            var hideControllerTimer = null;
            var trackWidth = parseInt(getStyle($track, 'width'), 10);
            var progressWidth = $progress.offsetWidth;
            // var canvas =  null;
            // if (this.canvas) {
            //     canvas = this.$canvas.get(0).getContext('2d');
            //     player.style['object-position']= '0px 0px';
            // }

            function showAndHideController() {
                if (hideControllerTimer) {
                    clearTimeout(hideControllerTimer);
                }
                $controller.style.visibility = 'visible';
                hideControllerTimer = setTimeout(function () {
                    $controller.style.visibility = 'hidden';
                }, INTERVAL_HIDE_CONTROLLER);
            }
            showAndHideController();

            function pauseShow() {
                //            $video.hide();
                //            $poster.show();
                playBtn.classList.remove('loading'); 
                playBtn.show();
                $playControl.classList.remove('pause');
            }
            function playShow() {
                // $poster.hide();
                playBtn.classList.add('loading');
                // if (!this.canvas) {
                //     $video.show();
                // }

                player.play();
                $playControl.classList.add('pause');
            }

            function getX(e) {
                var pageX = getPageX(e);
                if (!pageX) {
                    return -1;
                }
                var parentOffsetLeft = $progress.offsetLeft;
                //or $(this).offset(); if you really just want the current element's offset
                var relX = pageX - parentOffsetLeft;
                return relX;
            }

            function playToGivenPosition(e) {
                if (!player.duration) {
                    return;
                }
                var relX = getX(e);
                if (relX === -1) {
                    return;
                }
                var percentage = relX / progressWidth;
                player.currentTime = (player.duration * percentage);
                setWidth($played, (percentage * 100) + '%');
                _self._debugLog('playToGivenPosition', relX, progressWidth, percentage, player.duration, player.currentTime);
                // player.play();
            }
            function showLoadedPosition() {
                const buffered = player.buffered;
                if (buffered && buffered.length > 0) {
                    var buf = buffered.end(buffered.length - 1);
                    var percentage = buf / player.duration;
                    if (progressWidth * percentage > trackWidth) {
                        setWidth($loaded, percentage * 100 + '%');
                    }
                }
            }
            // function draw() {
            //     if(player.paused || player.ended) {
            //         return false;
            //     } 
            //     // First, draw it into the backing canvas
            //     canvas.drawImage(player,0,0,player.clientWidth,player.clientHeight);
            //     // Start over!
            //     setTimeout(function(){ draw(); }, 0);
            // }

            playBtn.addEventListener('click', function () {
                playShow();
            }, false);
            $leftControlArea.addEventListener('click', function () {
                if ($playControl.classList.contains('pause')) {//当前操作为暂停
                    player.pause();
                } else {//当前操作为播放
                    playShow();
                }
            }, false);


            player.addEventListener('error', function () {
                _self._debugLog('emit error events');
                pauseShow();
            }, false);


            player.addEventListener('pause', function () {
                _self._debugLog('emit pause events');
                pauseShow();
            }, false);

            player.addEventListener('timeupdate', function () {
                var currentTime = player.currentTime;
                if (!currentTime) {
                    return;
                }
                // $poster.hide();
                playBtn.hide();
                $playControl.classList.add('pause');
                // if (!_self.canvas) {
                //     $video.show();
                // }


                $timePassed.innerText = (secondsToString(currentTime));
                var percentage = (currentTime / player.duration);
                if (progressWidth * percentage > trackWidth) {
                    setWidth($played, (percentage * 100) + '%');
                } else {
                    setWidth($played, trackWidth);
                }
            }, false);
            player.addEventListener('loadedmetadata', function () {
                if (player.duration) {
                    $timeTotal.innerText = (secondsToString(player.duration));
                }
            });
            player.addEventListener('durationchange', function () {
                if (player.duration) {
                    $timeTotal.innerText = (secondsToString(player.duration));
                }
            }, false);
            player.addEventListener('loadeddata', function () {
                showLoadedPosition();
            }, false);
            player.addEventListener('canplaythrough', function () {
                showLoadedPosition();
            }, false);
            player.addEventListener('progress', function () {
                showLoadedPosition();
            });
            // if (this.canvas) {
            //     player.addEventListener('play',function() {
            //         draw();
            //     },false);
            // }

            $progress.addEventListener('click', function (e) {//进度条点击
                playToGivenPosition(e);
            });
            addListenerMulti($progress, 'mousedown touchstart', function (e) {
                showAndHideController();
                e.preventDefault();
            });
            addListenerMulti($progress, 'mouseup touchend', function (e) {
                playToGivenPosition(e);
            });
            addListenerMulti(this, 'mouseenter touchstart', function () {//触摸播放器，显示进度条
                showAndHideController(); _self._debugLog('player touch');
            });

            window.addEventListener('resize',function() {
                progressWidth = $progress.offsetWidth;
            }, false);
            $fullscreenArea.addEventListener('click', function () {
                _self._toggleFullScreen();
            }, false);
        }
    }
    customElements.define('whyun-player', WhyunPlayer);
})();